packages/agent/src/canisterStatus/index.ts

```diff
@@ -87,8 +87,8 @@ export const request = async (options: {
         const response = await agent.readState(canisterId, {
           paths: [encodedPaths[index]],
         });
-        const cert = new Certificate(response, agent);
-        const verified = await cert.verify();
+        const cert = new Certificate(response.certificate, agent.fetchRootKey());
+        const verified = await cert.verify(canisterId);
```

Comment:
quick check here - this changes the interface of `verify`. We should bump agent-js to the next minor version instead of the next patch version.

also, is there anywhere else in the codebase that calls `verify()` without arguments?

Reply:
I have the same concern. Not all certificates contain a path with canister_id. How do we verify the certificate if we only want to read non canister related paths, e.g. `/time`, `/subnet/` and `/request_status`?

Reply:
That's a good point - I'll bump the version up. We could also consider rejigging the type of `verify` (or getting rid of it) - see the answer to your next comment.

For the other point (other places calling `verify`), I'd hope that the Typescript type system is sound enough to catch this? However, I did miss one place, revealed by the failed PR tests (I didn't check these, my bad), in `e2e`, because I never called the TS compiler on that package. Let's discuss with Kyle to add a command that builds the whole monorepo in the right order and catches all such errors? I tried to do this as we discussed a couple of weeks ago, but didn't get anywhere in an hour or so that I spent on it.

Reply:
For @chenyan-dfinity 's point: any call to `read_state` (including those to get time, subnet and request status) has to have an effective canister ID, and the scope of the certificate must be checked against this ID:
https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-read-state
Is there another way to obtain such a certificate?

Reply:
> So we need to at least disable the check for the provisional API.

I think the interface spec is actually wrong on this, as it says that any canister is an effective canister ID for `provisional_create_canister_with_cycles`, but then also requires that the delegation is checked for the call to `read_state`. I think we should disable the check for the case when `aaaaa-aa` is the effective canister ID, as per the spec it can only be the effective canister ID for the `provisional_create_canister_with_cycles` call. I'll discuss this with Jens once he's back from vacation.

> Also, can this effective_canister_id be a member of the Certificate class, so that we don't have to pass the canister id when calling verify? This also makes it a non-breaking change. We can disable the check if the effective_canister_id is missing.

Since we're fixing a security flaw, I think we have to make a breaking change (callers that don't change the call would stay vulnerable). After discussing with Jason, since it's a breaking change already, I've changed the API to perform the verification immediately in the constructor - we couldn't find a use case where you'd want to delay the verification (do you maybe see one?). Also, I unified the error reporting to always use exceptions, instead of a mixture of exceptions and Booleans as was previously the case for `verify`

Reply:
> I think the interface spec is actually wrong on this
I think the spec is fine. We just need to disable the check when the effective canister id is `aaaaa-aa`. agent-rs does the same thing: https://github.com/dfinity/agent-rs/blob/main/ic-agent/src/agent/mod.rs#L1214

> Since we're fixing a security flaw, I think we have to make a breaking change (callers that don't change the call would stay vulnerable)

Not sure it has to be a breaking change. Callers that didn't upgrade agent-js to the fixed version stays vulnerable. It doesn't have to be a breaking change to fix a vulnerability.

> I've changed the API to perform the verification immediately in the constructor - we couldn't find a use case where you'd want to delay the verification (do you maybe see one?).

There might be a use case that the caller doesn't want to do a verification. If we verify in the constructor, it can be harder to skip verification? The problem is that bls verify is written in Wasm, some platforms don't support wasm execution, e.g. mobile platform, chrome plugins. In those environments, they either have to use a slower JS implementation or skip the verification completely. I know a grant awardee is writing a Chrome plugin to inspect all IC traffic, and cannot run verify in the plugin setting. Since it's a debugging tool, it's not too bad to disable the verification.